### ISSUE TITLE
ReQL optimizer

### ISSUE THREAD

**Issue by coffeemug** [rethinkdb/rethinkdb#4150]
When I talk to users about optimization, I get a 50/50 split:

- Half the people love explicit indexes. They like complete control of the execution and guarantees about asymptotic performance in production.
- Half the people hate explicit indexes, because for certain workloads using them turns out to be really, really hard.

For example, many users store time-series data (say, a log of all customer purchases over time), and give analysts a UI where they can filter and sort on random fields. An analyst might say:

- give me all the purchases where product type is shoes
- purchased after 2014
- sorted by price

Since the queries are essentially generated dynamically (the analyst toggles things in a custom UI to get what they want), it's *very* hard to write ReQL code that's executed efficiently. Even if you know the query ahead of time, it's hard to make it run efficiently. Should I use the product type index, the datetime index, or the price index? Which one performs better?

This is an extremely common use case.

I think the philosophy of explicit index specification is sound, and I think a traditional approach of writing queries and having the database optimize them behind the scenes is (a) hard to implement and (b) bad for live, realtime systems. I'd like to propose an alternative that satisfies the following goals:

- Allows us to maintain the philosophy of explicit index specification
- Is reasonably simple to implement
- Makes use cases like the one above dramatically easier to implement efficiently

```py
r.begin_optimizer()
 .table('purchases')
 .get_all('shoes', index='product_type')
 .between(r.time(2014, 1, 1), r.maxval, index='purchase_date')
 .order_by(index='price')
 .end_optimizer()
 .whatever_you_want()
```

The commands within the `begin/end_optimizer` block would have the following restrictions only allow a whitelist of indexable commands (i.e. `get_all`, `between`, etc.) Non-indexable commands and transformations aren't allowed (i.e. no `map`, `filter`, etc.) The optimizer can reorder commands within the block, and pick the indexes it believes are most valuable (i.e. it can choose not to use a particular index within the block).

It could be implemented either in a traditional way (by collecting statistics), or in a more creative way (e.g. run multiple versions of the query in parallel, kill the queries once the fastest one returns, and cache the information for a while).

This gives the following advantages:

- Fixes the most common use case where the optimizer is needed
- Is much easier to implement than a traditional optimizer since it's more limited, doesn't require parsing the AST within `filter`, etc.
- Allows us to maintain explicit indexing philosophy

/cc @mlucy

**Comment by thelinuxlich** [comment]
+1 for this, you've described my use case: the client feeds whatever data they want and after that they want to segment it into visitor profiles and after that there is a task that continuously search for users that fit these profiles and trigger custom actions(like mailing) on them

**Comment by mlucy** [comment]
I would much prefer:
```rb
r.optimize {
  ...
}
```
to:
```rb
r.begin_optimize()...end_optimize()
```

The fact that we didn't do that with `group/ungroup` has been biting us in the ass ever since, and in this case we're going to need to do syntax examination anyway so the argument about types doesn't really apply.

As for the actual meat of the proposal, this is definitely easier than other things we could do and solves a real use case without introducing too much magic.  I'd like to think about it more, but it's actually a pretty exciting proposal!

**Comment by thelinuxlich** [comment]
After this is implemented, RethinkDB will be a big choice for actionable analytics!

**Comment by coffeemug** [comment]
> ```rb
r.optimize {
  ...
}
```

I'd prefer that too, but how would you do that given that we don't allow binding streams to lambda arguments? Also, this doesn't *have* to be consistent with `group`, but it would be very nice. It might be worth refactoring `group` too if we figure out a way to do this well.

**Comment by marshall007** [comment]
If I'm not mistaken indexed operations can only happen in the first term (and possibly subsequent terms in the case of `order_by/between`) following `r.table`. Given that, is there any reason why the developer needs to define the portion of the query that should be subjected to optimization?

**Comment by marshall007** [comment]
> ... or in a more creative way (e.g. run multiple versions of the query in parallel, kill the queries once the fastest one returns, and cache the information for a while).

I realize you left `filter` out of this proposal for simplicity. However, if we go with the second option (quoted above), optimizing `r.table.filter` becomes super powerful so I couldn't help but mention it. Consider `r.table(...).filter({ foo: ..., bar: ... })` where `foo` and `bar` are both indexed. It's impossible to know at run-time which of the following is going to be faster, so executing both and running the faster one on subsequent requests would be a huge win!

```js
r.table(...).getAll(..., { index: 'foo' }).filter({ bar: ... })
r.table(...).getAll(..., { index: 'bar' }).filter({ foo: ... })
```

In one of our apps we currently do this transformation manually, but it's a naive approach. We have a hard-coded list of fields that we know are indexed and we do the `get_all` on the first one we come across then filter by everything else.

**Comment by danielmewes** [comment]
@marshall007 `filter` is tricky, since it doesn't allow specifying an index.

Part of the advantages of this proposal is that the query still explicitly defined the indexes that can be used. The difference is that with `optimize` usage of those indexes is made optional, so the server can pick which one works best.

In your example, you could simply rewrite the `filter({foo: ..., bar: ...})` to `getAll(..., {index: "foo"}).getAll(..., {index: "bar"})` and with enabled optimization RethinkDB would pick one of the two `getAll`+`filter` options that you mentioned (or execute both in parallel or something).

**Comment by danielmewes** [comment]
> Given that, is there any reason why the developer needs to define the portion of the query that should be subjected to optimization?

The idea is that optimize allows the terms between it to be re-ordered, so any of them could become the first term.

Having `optimize` as a run opt arg is an interesting alternative.

We could have the server automatically identify subregions of the query where re-ordering is possible, basically any subquery of the form `table.(between|orderBy|getAll)*`.
The advantage is that it's easier to use and nicer to read. The disadvantage is that it doesn't allow you to limit the optimization to certain subequeries. For example you might have a query
```js
r.table('t1').getAll(..., {index: "foo"}).getAll(..., {index: "bar"}).merge({
    other: r.table('t2').getAll(..., {index: "doo"}).coerceTo("ARRAY")
})
```

and you might want to have a guarantee that the query on `t2` is always executed as specified, while you want RethinkDB to pick the execution path for the part that's on `t1`.
I'm unsure how often this degree of control will actually be useful.

**Comment by mlucy** [comment]
@coffeemug -- we'll have to be transforming the source code anyway.  `.get_all` returns a type that you can't call e.g. `.between` on, so if a user writes `.get_all.between` our type system is already being violated and we'll have to rewrite it to something else.  As long as we're rewriting code anyway, it isn't that much more work to rewrite `optimize` with a lambda to something else as well, preventing us from trying to store a stream in a variable.  (It might still be confusing to users, though, which would be a reason not to do it.)

**Comment by timmaxw** [comment]
I like that this proposal involves a limited and well-defined set of optimizations. If `optimize` were a global optarg, it wouldn't be clear what parts of the query were getting optimized or how.

I'm a little uncomfortable with the idea of allowing operations inside `optimize` that we don't allow outside of `optimize`, like `.get_all.between`.

**Comment by marshall007** [comment]
@danielmewes that makes sense. I didn't read @coffeemug's example carefully enough. I assumed he was chaining the `getAll/between/orderBy` all on the same index (i.e. that it was still a "valid" ReQL command in today's world) and the only thing `optimize` was going to do was decide how best to order the terms.

The fact that it also decides which indices to use makes what I suggested less important. We were actually able to implement `filter` optimizations in `thinky` ([see PR](https://github.com/neumino/thinky/pull/126)) because we know about the indices on a model without having to query the table for them. However, I actually prefer the explicitly chained `getAll`s in your example over what I suggested because it makes things more obvious.

**Comment by marshall007** [comment]
> It could be implemented either in a traditional way (by collecting statistics), or in a more creative way (e.g. run multiple versions of the query in parallel, kill the queries once the fastest one returns, and cache the information for a while).

Regardless of how this piece is implemented I'd like to see the relevant stats exposed in the `stats` table. In the "creative way" case showing # of cache hits/misses and how often each optimized variation of the query gets executed would be very useful.

It would also be nice to be able to fine-tune the optimization cache timeout (assuming this doesn't happen automatically) and having this info available in the `stats` table will help in making informed decisions about what those values should be.

**Comment by danielmewes** [comment]
I agree that insight into and control over the optimization stats would be great.
What that's going to look like exactly and where it will be exposed is going to depend on which implementation we're going to chose I think.

**Comment by internalfx** [comment]
Why do I get the feeling that RethinkDB complexity is starting to spin out of control....

> Ahhh, the days when you could just create an index and **BAM** everything was faster.

Why do I feel like I'm the only one that thinks work should revolve around filter?
That the picture of success looks like filter (and only filter)?

Want to offer more power to control freaks? great. But all a user should **need** is filter.

Even most ORMs get this.

***

**User** Hi database! I'd like a table.
```
r.table('stuff')
```
**Database** Here's a table!
**User** uh...I actually don't want everything.
**Database** Oh, you want less than a table? Great! send me a filter.
**User** OKAY!
```
r.table('stuff').filter({idontwantthis: true})
```
**Database** Here you go! All the stuff you don't want is missing!
**User** Sweet! What took so long?
**Database** Oh I scanned the whole table. You need an index.
**User** OKAY!
```
r.table('stuff').filter({idontwantthis: true})
```
**Database** Here you go! All the stuff you don't want is missing!
**User** Sweet! What took so long?
**Database** Oh I scanned the whole table.
**User** Why? I made an index.
**Database** Well, you didn't tell me to use it.
**User** OKAY!
```
r.table('stuff').getAll(true, {'index': 'idontwantthis'})
```
**Database** Here you go! All the stuff you don't want is missing!
**User** Sweet! That was way faster!
**User** Okay, now I need all the stuff that I don't want, and my user is asking for only red items.
**Database** Use 2 filters.
**User** Won't that be slow?
**Database** Oh yeah, better go grab a frosty cold beverage.
**Database** You could use an index for one of the parameters, and filter the remaining.
**User** Hmm.. Which one should I use the index for?
**Database** How should I know? I'm just a database.
**Database** Well, there is a *special* trick you can use...
**User** You could tell me that you want it optimized...
***
> Isn't it kinda backwards to have it take *more* code to do it the easy way, and *less* code to do it the hard way?

Shouldn't it be assumed that I want my filter to run fast?



### Simplicity is the ultimate sophistication.

**Comment by thelinuxlich** [comment]
@internalfx has a point :D

**Comment by internalfx** [comment]
I'm being hard on the devs, but I hope it's taken from the angle that I wan't RethinkDB to be amazing.

If they could achieve the concept that filter was able to intelligently execute a query. So that it could be chained, yet still executed quickly. Then I think RethinkDB would reach a point where even non-programmers could get *some* use of it.

And, it would better mesh with the software that is created around it. As a dev, I could offer users options for filtering the data they work with without worrying about a **bad** performance edge case.

It would become so much more accessible...

**Comment by internalfx** [comment]
Essentially, in the above use case where filters are chained...

The solution that worked best for me, was to use ElasticSearch. Somehow, it's fast no matter what.

It is certainly outside the scope of the project. But it would be cool if I didn't need ElasticSearch for that. It seems like RethinkDB should be able to do this. Although, as I found out myself in my own *meager* attempt at optimization, it's complicated.

**Comment by mlucy** [comment]
@internalfx -- your point about simplicity is well taken.  I have a few worries about implicit indexing by default, though, especially if it's implemented by examining `filter` predicates and trying our best on them.

* Indexed and non-indexed queries have such radically different performance characteristics that they're effectively different operations.  There's a rule of thumb that different operations should look different so that you know which one you're doing.  (Anecdotally, I've talked to some very smart people using systems that do implicit indexing, and a surprising number of them have no idea which parts of their queries are indexed and which aren't.  I realize this is hard to believe, but it's true, and in some cases (e.g. compound indexes) it's hard to blame them since it isn't 100% obvious in systems like SQL which operations a given compound index makes fast.)
* Systems that try to automatically optimize queries for people inevitably end up implementing important functionality that only the optimizer knows how to do, which leads to the common and lamentable scenario of people having to twist their code into knots in order to fool the optimizer into doing what they want.
* If you talk to people who have run systems that automatically pick indexes for them (or do other similarly invasive query optimizations), most of them have a story where some query was running perfectly fine, they upgraded their system by a minor version, and performance fell through the floor because the optimizer was silently making different decisions under the hood.  (This ties into the previous point, because they usually end up having to fix it by fiddling with their query to trick the optimizer into doing the right thing.)

Basically, I think choosing indexes is usually too important of a decision to be done automatically under the hood in all cases.  Optimizations that speed things up by 20%, sure, do them automatically, but indexes can turn 5 hour queries into 100ms queries, and if you don't know what the optimizer is doing or the optimizer starts doing something different between versions it can make your app totally unusable.

I think the case being discussed here where app developers are producing complicated queries based on user input is a special case, because if we didn't offer a way of having the database choose indexes they'd be forced to write their own hacky version of it.

(Note: I would feel better about making `filter` a lot smarter if our profiler were better.  One of the advantages of the current system is that you can look at a giant ReQL query and know which parts are indexed and which aren't.  If our profiler made it easy to determine the same information at a glance, that would be less important.)

**Comment by internalfx** [comment]
@mlucy  I'm not sure that optimizing filters automatically leads us to the situation you are describing.

I'm not advocating for removing the ability to be explicit about queries.

Are you developing enterprise systems that handle thousands of users a day?
Are you worried that the next RethinkDB update will bring your app to a crawl?

> Then you should be explicit about what indexes you wish to use in your queries.

Is your app small?
Are you lazy?
Is it not the end of the world if your app goes down?

> Then you could roll the dice and let our optimizer do it for you.

***

See the difference?

Its still totally easy and accessible to new users.
It's easy for small apps.

If you have a big enterprise app, then you will need to know what you are doing anyway. That's not the realm of inexperienced devs.

Forcing explicit indexes is RethinkDB making assumptions about the apps requirements.

**Comment by kofalt** [comment]
I applaud the forward motion on getting query optimization into RethinkDB. As an author of a framework whose usage of RDB is 100% dynamically generated, I feel this proposal is in the right direction - but will need some adjustment to be of joy rather than pain. 

To demonstrate, here is my rough code design (draft only!) for using this feature as proposed:

1. At program launch, consume a script-generated or RDB-api-provided set of indexes on each table.
1. Begin constructing a query. Add `begin_optimizer()`.
1. When adding each Term, perform the following:
  1. Cross-index the implied table and desired operation with set of known indexes.
  1. Add index optarg if there's a match.
  1. Place the new Term in a buffer for optimized or un-optimized Terms.
1. Unroll the optimized buffer, add `end_optimizer()`, unroll the un-optimized buffer.
1. Execute; monitor deployments for slow queries.

Advocates of explicit index use might be laughing right about now :) This is obviously not complete and does not handle AST nesting - say, branches. I really would rather not write clumsy AST-buffering logic to try & massage a query into an optimizer block whitelist. 

The reason for the query construction shenanigans to keep application logic from knowing about indexes. I assert that **maintaining an arbitrary many-to-many mapping in unrefactorable function call optargs is an absolute anathema for maintainability and sanity**. I want *no part* in writing an application whose performance utterly depends upon every developer and refactor correctly (manually!) correlating a 30-to-300 or 300-to-3000 index:query ratio.

One's team will inevitably make a mistake, and will probably result in a tool that can parse your application code to look for written (static) queries and cross-reference it with known indexes. I am reminded of laboriously adding type assertions in a dynamic language, lacking a compiler to do it automatically.

Adding or removing any index implies a potential code diff in every portion of an application's persistence layer. Contrast this difficult to validate, change-management nightmare with:

1. Construct a SQL statement.
1. Execute; monitor deployments for slow queries.

I think this gets into dev vs ops: correctness & performance are two different concerns. The developer has the most complete information for making a query correct; I assert that (in many cases) the database has the most complete information for making a query performant.

@mlucy has provided some excellent points on the flip side of this assertion. My primary response to his post would be that optimizers should offer the option of returning the result query: remove concerns about magic by exposing the resultant transform. I've used this exact behavior before to great effect in a traditional RDBMS. Your assertion that "choosing indexes is usually too important of a decision to be done automatically under the hood in all cases" is reasonable, and I completely understand wanting to support that scenario.

The good news is that I think RDB can be perfectly capable of supporting both opinions, and we needn't debate philosophy to accomplish our goals. Changes that would make my approach significantly easier - while leaving @coffeemug's original intentions largely intact - are, in rough order of scope:

1. Allow `begin_optimizer()` ... `end_optimizer()` to contain non-indexed Terms. API contract could be that any & all Terms inside the optimizer block get re-ordered. More complicated than the proposed mandated whitelist, as the reordering has to be AST-equivalent (nested Terms, branching).

1. Given the above, for most queries I would have most / all of the query in an optimizer block... at which point a query optarg `optimize=true` as shorthand starts to make sense.

1. As mentioned previously, it would be great to have a query optarg that returns the query actually executed after optimization. I feel this would go a long way towards alleviating hidden-magic concerns.

1. The golden pony would be something like #2356, with an optarg on the optimizer block (or the entire query) adding indexes to Terms that lack them. Doing this outside the database, in a 'userland' query construction library, is simply never going to be optimal. I realize this one sounds unlikely :)

Adding merely the first component would eliminate the increased complexity burden this proposal currently threatens. I have a full application happily churning along ignorant of indexes, which get added by a query construction library, and adding an RDB optimizer is highly desirable. 

Ideally, these proposed tweaks would allow for both usage types to gain increased database functionality.

**Comment by internalfx** [comment]
@kofalt
>I think this gets into dev vs ops: correctness & performance are two different concerns. The developer has the most complete information for making a query correct; I assert that (in many cases) the database has the most complete information for making a query performant.

I Absolutely Agree on that. The database is the right place to optimize.

But why do you want an optarg at all?

Why not optimize by default, So that the following applies to RethinkDB?

1. Construct a RethinkDB statement.
2. Execute; monitor deployments for slow queries.

**Comment by kofalt** [comment]
Largely motivated by a desire to worth within the goals of the RDB developers. It's clear that there are differing opinions on index usage semantics, but that doesn't preclude a productive & pragmatic discussion on satisfying both use cases.

I don't have strong opinions about default vs opt-in, and would be happy with either. My main concern is making sure RDB optimizer design does not make it more painful to write index-agnostic queries.

**Comment by mlucy** [comment]
@kofalt -- thanks for your detailed thoughts.  If allowing non-indexed terms in the optimize block would make your use case a lot easier, I don't see any reason why we couldn't do that (we'll have to support reordering anyway).  Likewise for adding a global optarg.  Adding a way to return the optimized query to the user sounds really useful, but might be outside the scope of this issue.

I think we should consider using a name aside from "optimize" for this.  When I see the word "optimize", I assume that we're going to take non-indexed operations and turn them into indexed operations, while this proposal (in its current form) does the opposite: it lets you provide lots of indexed operations, then chooses some of them to turn into non-indexed operations and reorder.  I think we could avoid a lot of confusion if we called it `allow_unused_indexes` or something.  This would also fix the problem of people skimming the docs, seeing that there's a global "optimize" optarg to `run`, and then setting it to `true` and being confused that their queries don't get faster.  (`allow_unused_indexes` may be a bad name, but I'd bet we can come up with a good one.)

---

As for the question of giving people the option to have ReQL rewrite their queries to use indexes (like @internalfx proposes), I'm not necessarily 100% opposed to that, but it's a much larger project than this because it involves writing a static analyzer that can e.g. go into `filter` queries and figure out how to rewrite them to use the available indexes in the most efficient way.  (I would probably be opposed to turning it on by default, but this may just be a situation where what I would want is genuinely out of sync with what most users would want, so I could be convinced that we should turn it on by default if everyone else thinks it's a good idea.)

**Comment by danielmewes** [comment]
I agree with what @mlucy wrote.

Allowing arbitrary terms and a global optarg sound like great improvements.

In fact I think we should consider having *only* a global optarg for allowing RethinkDB to ignore indexes. The `begin_optimizer` and `end_optimizer` terms allow for more flexiblity, but is that really necessary? (also compare my earlier comment on this: https://github.com/rethinkdb/rethinkdb/issues/4150#issuecomment-97933556 )

I think picking suitable indexes for operations automatically is a separate issue, and we should keep it separate. While the use cases overlap, it requires a big deal of additional implementation work as far as I can tell. Especially since many ReQL terms right now aren't designed with that in mind. Many of them behave slightly differently when used with an index versus without one.

Allowing the database to ignore indexes and re-order terms is by itself useful, and I think we shouldn't delay shipping that feature by trying to make index usage completely automatic.

(Also @kofalt sorry for not replying earlier. Thanks also from my side for your great feedback.)

**Comment by marshall007** [comment]
Have you guys thought about how the optimizer is going to disambiguate between indices that point to a field vs. the ones that are computed? Given @coffeemug's example, let's say the `price` index is actually defined as `r.row('list_price').mul(r.row('discount'))`. In cases like this the optimizer would have no choice but to do the `.order_by(index='price')` first, correct? Since it wouldn't have access to the secondary index function used to compute those values.

**Comment by danielmewes** [comment]
@marshall007 I think the optimizer could rewrite `order_by(index='price')` to `order_by(index_fun)` where `index_fun` would be the index function by which the `price` index is defined.
While we don't currently expose the stored index function through ReQL, we could definitely give the optimizer a way of loading it.

**Comment by kofalt** [comment]
These updates sound very positive.

As I understand it @mlucy + @danielmewes have endorsed proposals 1 and/or 2 from my comment.
If either of those come to fruition I will be greatly satisfied with the result. 

The distinction between "fiddling with specific index usage" versus "optimizing" a query makes sense, and I appreciate the later being much larger in scope - attacking the former first is understandable.

My approach to RDB is currently at a distinct disadvantage due to the way I incrementally modify a query, thus the shenanigans proposed in my feedback above. My ambition is to eventually replace (and open-source!) the current implementation with an AST-walker as proposed on [my comment on #2356](https://github.com/rethinkdb/rethinkdb/issues/2356#issuecomment-68220306). Both the current & possible future approaches will be significantly easier with the proposed changes. Those interested in contributing to such a project (in Go) should contact me privately!

In summary: Allowing arbitrary terms and/or a global optarg would solve my qualms with this feature.

**Comment by gmethvin** [comment]
I agree with @kofalt and @internalfx.

To me, manually ordering terms and explicitly defining which indexes are used is an advanced feature, something you would only want to do when you have a complex enterprise app that needs to have more predictable performance. I think a lot of users expect the database to do the "best" thing by default.

**Comment by Morgul** [comment]
So, having read through the first 2/3rds of this issue and skimmed the rest, I feel amy use case could be beneficial as a talking point.

We have roughly 2 million documents, and wish to store their metadata in RethinkDB. These documents are generated by different business operations, and each has a defined set of business rules for what their metadata must contain, and a less well defined set of what their metadata _may_ contain. (Ex: all insurance forms must have the policy number, but they may or may not also have the social security number, and if they were generated in this special way, they can also contain the date the form was printed.)

I have been tasked with creating a single service that allows for document retrieval; details are unimportant, but the requirement is that it be a RESTful site, which allows querying of this meta data. This means when a user wants to search for documents, my query amounts to:

```javascript
r.db('document_store').table('metadata').filter(queryParams)
```

How, with this proposal, would I do this efficiently? In a normal RDBMS, I would add secondary indexes on the most commonly used fields, and let it optimize my `where` clause. What I would prefer is, essentially:

```javascript
r.db('document_store').table('metadata').filter(queryParams, { use_indexes: true })
```

Given the discussion, if the solution is instead to write:

```javascript
r.db('document_store').table('metadata')
    .getAll(555555, { index: 'policyNum' })
    .getAll(999, { index: 'formCode' })
    .getAll(...)
    ...
```

That meets my needs, since I can write a function to programmatically generate that query, based on user input. It's still not _convenient_ (which I think is worth a consideration), but it is _acceptable_. 

Frankly, if (in the optimizer case) `.getAll('bar', { index: 'foo' })` === `.filter({foo: bar})`, why should I have to write the more clunky syntax, and chain multiple `.getAll` calls? If all you want is an explicit specification of what index should be used, why not do this: `.filter({policyNum: { value: 555555, index: 'policyNum', formCode: { value: 999, index: 'formCode' } } })`? It's less repetitive, and is also explicit about the indexes to use. (It annoys me if my indexes are simple, therefore the field name is repeated, but that's minor.)

What's not clear to me, however, is if my last example matches what is being proposed, or not? If not, how do I handle my use case?

**Comment by danielmewes** [comment]
@Morgul thank you for sharing your thoughts on this. The `getAll` solution that you mention is what's currently being proposed.

The main advantages of this over enabling general `filter` terms to use indexes transparently are:
* It's easier for us to implement, because we don't have to analyze the `filter` function. We could restrict this to work only for simple predicates of the form `filter({field: value, ...})`, but that would make it impossible to use this for `between` predicates for example.
* Furthermore, it's hard to deduce which indexes can be used for a specific `filter` predicate. If all indexes are on individual fields this is relatively simple, but we allow indexes on all sorts of functions and combinations of fields. Again, we could restrict the optimizer to make use of only simple indexes, but it would also limit its applications.

Your syntax of specifying the `index` value "inline" in the `filter` predicate is an interesting idea, but I think it's problematic in practice because we cannot distinguish whether the predicate `filter({field: {value: 'a', filter: 'field'}})` means that you want to filter for documents with `field = 'a'`using the index `field`, or if you want to filter for documents where `field` has the actual object `{value: 'a', filter: 'field'}` as its value.

One thing that might work is to add support for the more explicit `.getAll(..., {index: ...}).getAll(..., {index: ...})...` syntax, and additionally supporting automatic use of simple (single-field, and possibly compound) indexes for simple filter predicates of the form `filter({field1: val1, field2: val2, ...}) .
That way the feature would be really convenient to use in the case you describe, while users who need more complex predicates (e.g. filtering by ranges using `between`) or who want to use non-trivial indexes can still use the more explicit syntax.
Furthermore we could implement and test the explicit syntax first, and add the `filter` shortcut later.

**Comment by Morgul** [comment]
@danielmewes Thanks for the clarification. To address a few more points:

* **Analyzing the filter** - I definitely understand the complexity here. I think it might be acceptable to roll out optimization in stages; a first pass that will work for simple predicates in `filter`, and then as time and importance allows, this gets expanded until eventually it's a full blown analysis. Out of curiosity, when you say _"but that would make it impossible to use this for `between` predicates for example"_, what do you mean by 'impossible'? Can you expand a little bit more on that?

* **Picking indexes** - I think there's a feeling that optimization needs to 90% correct to be worthwhile; when in reality I think anything over 50% is a win. (Technically, anything over 0% is a win, but I digress...) So, the simplest thing would be to say, "When filter uses secondary indexes, it will only use simple indexes matching the keys of the predicate." I feel that's worthwhile, and addresses several use cases, while being 'no worse' than currently. (Caveat: I assuming that `filter` only uses indexes when explicitly told to do so, by an option. This is in keeping with the desires of the current discussion, and prevents wrong or failed queries because you named a compound index the same as a field you're trying to filter on.)

Regarding how to tell filter what indexes to use, here's an alternate syntax that could work:

```javascript
r.db('document_store')
    .table('metadata')
    .filter({ policy: 555555, form: 999 },
        { index_map: { policy: 'crazyIndexOfDoom', form: 'form' }
    })
```

It's a bit repetitive, but is trivially generated in code, and has a nice explicitness to it; it maintains the "You know how optimized each piece of the query is" property that's desired. The one addition I would request is an option: `use_simple_indexes` that would attempt to use simple indexes for each key of the filter, and not finding one, would work as filter currently does. This keeps filter explicit, and (unless I miss something) keeps the optimizer simple, as proposed. (Note: I'm assuming that telling filter to use secondary indexes is _actually_ telling it to use the simple optimizer discussed here; my limited understanding is that just blindly using secondary indexes wouldn't be as desirable as attempting to optimize their usage. However, if that's not the case, then maybe we don't need an optimizer?)

As for your last proposal for supporting both syntaxes; sure. I suspect we could create a more elegant syntax (perhaps like what I proposed, or something different), but the core of what you suggest is very much in line with my thoughts.

**Comment by eldritch-fossicker** [comment]
As a DBA working with both SQL Server and PostgreSQL I get the frustration of working with a query optimize. I also understand the the complexity of code that goes into writing such a beast.  My concerns about no or a limited optimizer are:

*  Metadata availability  #3231, someone/something has to optimize the query. If its me or code from the dev I work with we need a systematic and authoritative way to get the information.
* If there are limits on the order and/or the number of index enable operations (I hope there is not) it would be nice for it to very clear in the ReQL language definition because this could be a deal breaker.  * some kind of dry-run tool immensely helpfully in learning ReQL and verifying these hand-built plans

**Comment by timshannon** [comment]
I completely agree @eldritch-fossicker .  SQL Servers's query optimizer seems almost non-deterministic at times.  We've had queries that run fine in our testing environments, and ridiculously slow in Production because the query optimizer came up with a different plan.

With ReQL in it's current state you explicityly tell rethinkdb how you want it to retrieve your data.  I'm fine with the addition of a query optimizer, just so we're not forced to use it (which sounds like the proposed plan).

**Comment by internalfx** [comment]
agree with @Morgul.

Indexes using simple keys is all I have in use right now. But it would be a huge convenience if they were used automatically.

**Comment by internalfx** [comment]
Glad to see this issue getting traction again.

**Comment by Zolmeister** [comment]
+1 for simple filter optimization
Also, I feel this is important (as with #809), and really ought to be scheduled for an upcoming milestone.

**Comment by coffeemug** [comment]
> "but that would make it impossible to use this for between predicates for example", what do you mean by 'impossible'?

@Morgul -- what @danielmewes meant here is that if we restrict the optimizer to terms like `filter({a: b})`, the optimizer wouldn't be able to optimize queries of the type `a < b < c` (where as if we went with the original proposal, we'd be able to optimize both). I actually like your syntax (though we'd have to figure out how to optimize things like "give me every user with age greater than 30").

For anyone watching this issue, this is still a bit far away, but will definitely get done. Without committing to a specific ETA, I think we can get it into RethinkDB 2.4 or 2.5. (I realize that's inconvenient, and I'd *love* to do this earlier, but Rethink has been picking up popularity, and we have a ton of very valid feature requests from all kinds of different users and unfortunately limited engineering throughput capacity)

**Comment by kofalt** [comment]
:+1: @coffeemug. Great to hear it!

**Comment by internalfx** [comment]
@coffeemug Any updates on this? This single feature would likely eliminate most of my performance problems...

**Comment by coffeemug** [comment]
No updates yet, but qualitatively I'd be shocked if this got done in 2015 and if it didn't get done in 2016.

**Comment by gamesbrainiac** [comment]
So, this is basically what i want to say. Have a flag on `conn` if you want a certain query to be optimized, this way you can just stick it into the `run` method of the query and this will work across drivers and won’t interfere with what hard core optimizers want, which is to hand pick their indexes. In the mean time, improve the optimizer so that at one point you won’t need to hand pick your indexes anymore.

However, this raises an issue that what if the optimizer does something unexpected. Well, if we can build a tool comparing average query time, we can help identify the queries that need hand picked indexes to run optimally.

With any optimizer, things will always be a work in progress, but progress will be made with efficiency. With this approach, we can pave the way for the beginnings of an optimizer that gets better with each iteration while leaving those who need highly optimized queries to get their jobs done, the option to do so.

**Comment by kofalt** [comment]
@gamesbrainiac You might be interested in this https://github.com/rethinkdb/rethinkdb/issues/4150#issuecomment-100543240 upthread, where I laid out a few proposals. The comment itself is pretty long; most relevant parts are near the end.

Also of interest are RDB staff comments following thereafter, regarding my first & second proposal.

**Comment by hiddentao** [comment]
I just hit this issue myself. It's 2016 now - is is on the roadmap for this year?

**Comment by coffeemug** [comment]
@hiddentao -- it isn't at the moment because of some higher priority projects. We're shipping something we're really excited about in April -- we'll reevaluate after that! (sorry to keep you waiting)

**Comment by saikat** [comment]
RethinkDB 2.3 looks amazing (assume that's 'something' in April that you all were really excited about)!  Any thoughts on if this will be added on to the roadmap now that 2.3 is out?  Sorry to pester!

**Comment by danielmewes** [comment]
@saikat We haven't fully fixed the set of features for 2.4 yet, but I think the primary candidates for that release are https://github.com/rethinkdb/rethinkdb/issues/3997 and https://github.com/rethinkdb/rethinkdb/issues/746 .

I think realistically we will get to this one (optimizer) for either 2.5 or 2.6.

**Comment by Pupix** [comment]
Hello,

Are there any news on this topic since the new governance?

I am on the same boat as @internalfx and @Morgul , I understand that hand crafted queries that you can look at and spot usage of indices is nice to have, however dynamic queries are common in my projects, where users have control of the displayed data and it's either painful to implement or slow.

Here's an example use case:
Imagine a tabular layout, where a user has a search bar that allows you to write multiple fields (name, email, status, role, department and so on) and then this user can order the results on an arbitrary field.

The optimal solution in my eyes (as the developer of the application and user of the database) would be to create indexes on fields that are publicly exposed, have the user select their filters and create a request from that data, let's say a GET, then the server can convert the query parameters in a `.filter(...)` (using object or function) and an `orderBy(...)`. At this point the DB can take the wheel and make my query as fast as possible.

Same thing could be said about data that needs to be displayed on graphs, where a user wants to see how much of X, with these characteristics, has been sold or whatever, in a time frame.

If there are clean solutions for this kind of use case already, let me know.

A side note for @mlucy:
The idea that implicit use of indices is bad because the optimizer could make different decisions based on version seems iffy to me. For all I know there could be regressions in the whole querying process when updating to a different version, and I'd be facing the same problem, where I have to stick with a version until I figure out where the problem is with the new one.

Cheers.
